### PR TITLE
chore: include start/endTime and duration in onEnd report callback

### LINKED ISSUE #23637: [Feature] - Add timestamp to Playwright Test Report
Hi Playwright team,

I've been making extensive use of your fantastic tool for running automated test suites. The tests are scheduled to run periodically, and they produce a playwright report that is hosted on Github Pages. This is an extremely convenient setup, but there's one feature I feel would greatly enhance the usefulness of the report – a timestamp.

As it stands, the reports generated do not have a timestamp, and that makes it challenging to discern when the tests were run or when the report was generated, especially when we're dealing with a large number of test runs and reports.

I think adding a timestamp feature to the report would greatly help in adding some context and timeline to the reports. This way, we could easily see when each report was generated, making it easier to track changes, identify patterns, and debug issues.

Given the open nature of this project, I am more than willing to help add this feature myself. However, I would appreciate some guidance on how best to approach this task. If you could provide me with some instructions or pointers to where the relevant files might be in the repo, it would be greatly appreciated. I'm not familiar with the codebase, so some directions to speed things up would be incredibly helpful.

Thank you in advance for your time and consideration.

### PR DESCRIPTION
**Pull request by pavelfeldman**
Fixes https://github.com/microsoft/playwright/issues/23637

### CODE REVIEW

File: packages/playwright-test/src/reporters/merge.ts [hunk 1]
```diff
@@ -260,16 +259,29 @@ function mergeConfigs(to: JsonConfig, from: JsonConfig): JsonConfig {
...
+    else if (shardResult.status === 'interrupted' && status !== 'failed' && status !== 'timedout')
+      status = 'interrupted';
+    startTime = Math.min(startTime, shardResult.startTime);
+    endTime = Math.max(endTime, shardResult.endTime);
+    duration += shardResult.duration;
   }
+  const result: JsonFullResult = {
+    status,
+    startTime,
+    endTime,
```
Comment (dgozman) [inline packages/playwright-test/src/reporters/merge.ts:282]: Perhaps we should set `endTime` to `now()`, but `startTime` to `endTime - duration`?

File: packages/playwright-test/src/reporters/merge.ts [hunk 2]
```diff
@@ -260,16 +259,29 @@ function mergeConfigs(to: JsonConfig, from: JsonConfig): JsonConfig {
...
-    else if (shardResult.status === 'interrupted' && result.status !== 'failed' && result.status !== 'timedout')
-      result.status = 'interrupted';
+      status = 'failed';
+    else if (shardResult.status === 'timedout' && status !== 'failed')
+      status = 'timedout';
+    else if (shardResult.status === 'interrupted' && status !== 'failed' && status !== 'timedout')
+      status = 'interrupted';
+    startTime = Math.min(startTime, shardResult.startTime);
+    endTime = Math.max(endTime, shardResult.endTime);
+    duration += shardResult.duration;
```
Comment (dgozman) [inline packages/playwright-test/src/reporters/merge.ts:277]: IIRC, merger already uses `metadata.totalTime` that we can remove with this change.

File: packages/playwright-test/src/reporters/html.ts [hunk 3]
```diff
@@ -274,6 +274,9 @@ class HtmlBuilder {
     }
     const htmlReport: HTMLReport = {
       metadata,
+      startTime: result.startTime.getTime(),
+      endTime: result.endTime.getTime(),
+      duration: result.duration,
       files: [...data.values()].map(e => e.testFileSummary),
       projectNames: projectSuites.map(r => r.project()!.name),
       stats: { ...[...data.values()].reduce((a, e) => addStats(a, e.testFileSummary.stats), emptyStats()), duration: metadata.totalTime }
```
Comment (dgozman) [inline packages/playwright-test/src/reporters/html.ts:282]: This line already includes `stats.duration`

File: docs/src/test-reporter-api/class-reporter.md [hunk 4]
```diff
@@ -113,6 +113,9 @@ Called after all tests have been run, or testing has been interrupted. Note that
 * since: v1.10
 - `result` <[Object]>
   - `status` <[FullStatus]<"passed"|"failed"|"timedout"|"interrupted">>
+  - `startTime` <[Date]>
+  - `endTime` <[Date]>
+  - `duration` <[int]>
```
Comment (dgozman) [inline docs/src/test-reporter-api/class-reporter.md:118]:
```suggestion
  - `duration` <[int]> Total testing time in milliseconds.
```

**Review by dgozman — APPROVED** [review]

File: packages/playwright-test/src/reporters/merge.ts (hunk 1 above)
Comment (pavelfeldman) [reply packages/playwright-test/src/reporters/merge.ts:282]: Can't do math on monotonic vs wall time.

File: packages/playwright-test/src/reporters/merge.ts (hunk 2 above)
Comment (pavelfeldman) [reply packages/playwright-test/src/reporters/merge.ts:277]: How do I make sure onEnd in the combined report contains the right duration?

File: docs/src/test-reporter-api/class-reporter.md [hunk 5]
```diff
@@ -113,6 +113,8 @@ Called after all tests have been run, or testing has been interrupted. Note that
 * since: v1.10
 - `result` <[Object]>
   - `status` <[FullStatus]<"passed"|"failed"|"timedout"|"interrupted">>
+  - `startTime` <[Date]>
+  - `duration` <[int]>
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply docs/src/test-reporter-api/class-reporter.md:117]: We document these in overrides for some reason.

File: packages/playwright-test/src/reporters/merge.ts [hunk 6]
```diff
@@ -252,24 +250,33 @@ function mergeConfigs(to: JsonConfig, from: JsonConfig): JsonConfig {
...
-      result.status = 'timedout';
-    else if (shardResult.status === 'interrupted' && result.status !== 'failed' && result.status !== 'timedout')
-      result.status = 'interrupted';
+      status = 'failed';
+    else if (shardResult.status === 'timedout' && status !== 'failed')
+      status = 'timedout';
+    else if (shardResult.status === 'interrupted' && status !== 'failed' && status !== 'timedout')
+      status = 'interrupted';
+    startTime = Math.min(startTime, shardResult.startTime);
+    duration += shardResult.duration;
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply packages/playwright-test/src/reporters/merge.ts:273]: You can't do math on non-monotonic times.

**Review by yury-s — APPROVED** [review]

**Pull request merged**